### DISCUSSION TITLE
h2 데이터베이스 DB_CLOSE_ON_EXIT 설정

### DISCUSSION THREAD

**Discussion by janeljs** [Sc0up/scoup-backend#137] [Troubleshooting docs]
~`DB_CLOSE_ON_EXIT=FALSE`으로 설정하면 안된다고 함~ &rarr; 잘못된 정보로 판정
```
spring.datasource.url: 'jdbc:h2:mem:testdb;DB_CLOSE_DELAY=-1'
```

- http://www.h2database.com/html/features.html#in_memory_databases

> Don't Close a Database when the VM Exits
By default, a database is closed when the last connection is closed. However, if it is never closed, the database is closed when the virtual machine exits normally, using a shutdown hook. In some situations, the database should not be closed in this case, for example because the database is still used at virtual machine shutdown (to store the shutdown process in the database for example). For those cases, the automatic closing of the database can be disabled in the database URL. The first connection (the one that is opening the database) needs to set the option in the database URL (it is not possible to change the setting afterwards). The database URL to disable database closing on exit is:
```
String url = "jdbc:h2:~/test;DB_CLOSE_ON_EXIT=FALSE";
```

**Comment by sallyjellyy** [comment]
[여기](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.sql)에 이런 내용이 있네요!
> If, for whatever reason, you do configure the connection URL for an embedded database, take care to ensure that the database’s automatic shutdown is disabled. If you use H2, you should use DB_CLOSE_ON_EXIT=FALSE to do so.

**Reply by janeljs** [reply]
제가 아래 답변만 보고 오독했는데

https://stackoverflow.com/questions/52480973/h2-in-memory-test-db-closed-despite-db-close-on-exit-false

>Use DB_CLOSE_DELAY only.
For in-memory database, you shouldn't use DB_CLOSE_ON_EXIT=FALSE: you should only use DB_CLOSE_DELAY=-1. See http://www.h2database.com/html/features.html#in_memory_databases
So you datasource should be:
spring.datasource.url: 'jdbc:h2:mem:testdb;DB_CLOSE_DELAY=-1'

**Reply by janeljs** [reply]
필자가 잘못 답변을 단 거고 h2 공식문서도 같은 입장을 취하고 있네요ㅎㅎ

> Don't Close a Database when the VM Exits
By default, a database is closed when the last connection is closed. However, if it is never closed, the database is closed when the virtual machine exits normally, using a shutdown hook. In some situations, the database should not be closed in this case, for example because the database is still used at virtual machine shutdown (to store the shutdown process in the database for example). For those cases, the automatic closing of the database can be disabled in the database URL. The first connection (the one that is opening the database) needs to set the option in the database URL (it is not possible to change the setting afterwards).
